hips/hip-9999.md

```diff
@@ -0,0 +1,79 @@
+- Third-party tooling to that provides recommendations on helm upgrades can query the cluster to get the repository URL to aid in this recommendation. See [nova](https://github.com/FairwindsOps/nova) as an example.
+- This metadata field could also populate output for the `helm ls` command in the future. That may be outside the scope of this particular feature.
+
+## Rationale
+
+In order to query for this in the cluster, the repository URL will need to be somewhere in the release object. The chart metadata portion of the release object already contains similar data and makes the most sense for this purpose.
+
+## Specification
+
+Nothing would change from the user's perspective. With some modifications in the current code it's possible to add this metadata into the release object transparently.
+
+The current idea for this implementation is to add a field in both the `ChartDownloader` and `Metadata` (in the `pkg/chart/metadata.go` file) structs to hold the repository URL. The Downloader is the logical place to derive the proper URL and can then that storage location can populate the Metadata field.
+
+More specifically, the URL is included in the local repository index file for a locally named repository. For instance, running `helm repo add [NAME] [URL]` would add the repository to the index file with the `[URL]` passed as the real location to download charts. In the helm code, the downloader reads this file to find the URL which is what gets bubbled up to the release object.
+
+If for any reason this data can not be gathered at install/upgrade time, then that particular field would be omitted from the release object metadata.
+
+The chart metadata object would have new fields that looked like so:
+
+```
+"repoURL": "https://charts.fairwinds.com/stable"
+"tarballDigest": "sha256:f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9"
+```
+
+And an OCI compliant repository would look like this (example is using ECR):
+
+```
+"repoURL": "oci://${ACCOUNT_NUMBER}.dkr.ecr.us-east-1.amazonaws.com/goldilocks"
+"tarballDigest": "sha256:f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9f9a9"
+```
+
+## Backwards compatibility
+
+There should be no impact to backwards compatibility.
+
+## Security implications
+
+A potential security issue would be private repository URLs becoming present in a cluster via this implementation. A potential fix for this would be to add a command line argument that would keep this field in the release object empty, but that would introduce something a user has to change in their workflow which may not be desirable. Another way would be to attempt a DNS match against a publicly available DNS server and if it's not found, then we do not populate the repository URL field.
+
+Another option is to ignore this potential downside. Unclear if this would truly be an issue or not. Would love to have feedback on why it would be bad for an internal repository URL to be stored in the cluster.
```

Comment:
As an additional (positive) security implementation, giving the `tarballDigest` would mean that utilities could independently verify that a chart has not been altered. However, a caveat should be noted here: There is no guarantee that `helm` will rebuild exactly the same tarball if presented exactly the same input _data_ because some input _metadata_ (timestamps, ordering) may change. So we wouldn't necessarily want to assume that just because a tarball has a different SHA, it's _data_ has changed and thus it is untrustworthy.

Since this HIP does not have a suggested implementation of how `tarballDigest` should be implemented in the client, I am fine with merely noting this issue under Security Implications and then letting implementors figure out how to cope.